Use of deprecated asyncio.coroutines.iscoroutine

Our implementation of `deprecated` currently uses `asyncio.coroutines.iscoroutine`, but that function is deprecated as of Python 3.14.

We fell over this with `typing-extensions-4.16.0rc1` and `pydantic`:

```bash
pip install --pre pydantic
python -W error -c "import pydantic"   # Our CI always raises warnings as errors to force handling warnings from upstream dependencies
```

```python
/usr/local/lib/python3.14/site-packages/ewokscore/task.py:25: in <module>
    from pydantic import ValidationError
/usr/local/lib/python3.14/site-packages/pydantic/__init__.py:5: in <module>
    from ._migration import getattr_migration
/usr/local/lib/python3.14/site-packages/pydantic/_migration.py:5: in <module>
    from pydantic.warnings import PydanticDeprecatedSince20
/usr/local/lib/python3.14/site-packages/pydantic/warnings.py:5: in <module>
    from .version import version_short
/usr/local/lib/python3.14/site-packages/pydantic/version.py:7: in <module>
    from pydantic_core import __version__ as __pydantic_core_version__
/usr/local/lib/python3.14/site-packages/pydantic_core/__init__.py:31: in <module>
    from .core_schema import CoreConfig, CoreSchema, CoreSchemaType, ErrorType
/usr/local/lib/python3.14/site-packages/pydantic_core/core_schema.py:4373: in <module>
    @deprecated('`field_before_validator_function` is deprecated, use `with_info_before_validator_function` instead.')
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.14/site-packages/typing_extensions.py:3142: in __call__
    if asyncio.coroutines.iscoroutinefunction(arg):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.14/asyncio/coroutines.py:23: in iscoroutinefunction
    warnings._deprecated("asyncio.iscoroutinefunction",
/usr/local/lib/python3.14/_py_warnings.py:830: in _deprecated
    _wm.warn(msg, DeprecationWarning, stacklevel=3)
E   DeprecationWarning: 'asyncio.iscoroutinefunction' is deprecated and slated for removal in Python 3.16; use inspect.iscoroutinefunction() instead
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -83,7 +83,7 @@ jobs:
           cd src
           python --version  # just to make sure we're running the right one
           # Run tests under coverage
-          python -m coverage run -m unittest test_typing_extensions.py
+          python -We -m coverage run -m unittest test_typing_extensions.py
           # Create xml file for Codecov
           coverage xml --rcfile=../pyproject.toml --fail-under=0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,8 @@
+# Unreleased
+
+- Avoid a `DeprecationWarning` when applying `deprecated` to a coroutine function on
+  Python 3.14.0.
+
 # Release 4.16.0rc1 (June 24, 2026)
 
 - Make `typing_extensions.TypeAliasType`'s `__module__` attribute writable.

--- a/src/typing_extensions.py
+++ b/src/typing_extensions.py
@@ -3130,7 +3130,6 @@ else:
                 __init_subclass__.__deprecated__ = msg
                 return arg
             elif callable(arg):
-                import asyncio.coroutines
                 import functools
                 import inspect
 
@@ -3139,11 +3138,13 @@ else:
                     warnings.warn(msg, category=category, stacklevel=stacklevel + 1)
                     return arg(*args, **kwargs)
 
-                if asyncio.coroutines.iscoroutinefunction(arg):
+                if inspect.iscoroutinefunction(arg):
                     # Breakpoint: https://github.com/python/cpython/pull/99247
                     if sys.version_info >= (3, 12):
                         wrapper = inspect.markcoroutinefunction(wrapper)
                     else:
+                        import asyncio.coroutines
+
                         wrapper._is_coroutine = asyncio.coroutines._is_coroutine
 
                 arg.__deprecated__ = wrapper.__deprecated__ = msg

